engine/verification/fetcher/engine.go

```diff
@@ -106,352 +101,206 @@ func (e *Engine) Done() <-chan struct{} {
-	header, err := headers.ByBlockID(blockID)
-	if err != nil {
-		return false, nil, fmt.Errorf("could not get block header by ID: %w", err)
-	}
-
-	lastSealed, err := state.Sealed().Head()
-	if err != nil {
-		return false, nil, fmt.Errorf("could not get last sealed: %w", err)
-	}
-
-	sealed := header.Height <= lastSealed.Height
-	return sealed, header, nil
-}
+// HandleChunkDataPack is called by the chunk requester module everytime a new requested chunk data pack arrives.
+// The chunks are supposed to be deduplicated by the requester.
+// So invocation of this method indicates arrival of a distinct requested chunk.
+func (e *Engine) HandleChunkDataPack(originID flow.Identifier, chunkDataPack *flow.ChunkDataPack, collection *flow.Collection) {
+	lg := e.log.With().
+		Hex("origin_id", logging.ID(originID)).
+		Hex("collection_id", logging.ID(collection.ID())).
+		Hex("chunk_id", logging.ID(chunkDataPack.ChunkID)).
+		Logger()
+	lg.Info().Msg("chunk data pack arrived")
 
-// return agrees and disagrees.
-// agrees are executors who made receipt with the same result as the given result id
-// disagrees are executors who made receipt with different result than the given result id
-func executorsOf(receipts []*flow.ExecutionReceipt, resultID flow.Identifier) ([]flow.Identifier, []flow.Identifier) {
-	agrees := make([]flow.Identifier, 0)
-	disagrees := make([]flow.Identifier, 0)
-	for _, receipt := range receipts {
-		executor := receipt.ExecutorID
-		if receipt.ExecutionResult.ID() == resultID {
-			agrees = append(agrees, executor)
-		} else {
-			disagrees = append(disagrees, executor)
-		}
+	// make sure we still need it
+	status, exists := e.pendingChunks.ByID(chunkDataPack.ChunkID)
+	if !exists {
+		lg.Debug().Msg("could not fetch pending status from mempool, dropping chunk data")
+		return
 	}
-	return agrees, disagrees
-}
-
-// onTimer runs periodically, it goes through all pending chunks, and fetches
-// its chunk data pack.
-// it also retries the chunk data request if the data hasn't been received
-// for a while.
-func (e *Engine) onTimer() {
-	allChunks := e.pendingChunks.All()
 
-	e.log.Debug().Int("total", len(allChunks)).Msg("start processing all pending pendingChunks")
+	lg = lg.With().
+		Uint64("chunk_index", status.Chunk.Index).
+		Hex("block_id", logging.ID(status.Chunk.BlockID)).
+		Hex("result_id", logging.ID(status.ExecutionResultID)).Logger()
 
-	sealed, err := e.state.Sealed().Head()
+	// make sure the chunk data pack is valid
+	err := e.validateChunkDataPack(status.Chunk, originID, chunkDataPack, collection)
 	if err != nil {
-		e.log.Error().Err(err).Msg("could not get last sealed block")
+		// TODO: this can be due to a byzantine behavior
+		lg.Error().Err(err).Msg("could not validate chunk data pack")
 		return
 	}
 
-	allExecutors, err := e.state.Final().Identities(filter.HasRole(flow.RoleExecution))
+	result, err := e.results.ByID(status.ExecutionResultID)
```

Comment:
The logic for the current implementation is as follows:

We have a chunk `status` per the assigned chunk of a result, and we keep all these statuses in memory, basically since they are ephemeral. Now if we replicate the result as a part of each `status`, we are adding a high degree of redundancy to the memory utilization, which may lead to a higher rate of out-of-memory bumping on the nodes. This is more likely to happen on the larger number of `worker`s. Considering that all assigned chunks of the same result will have a copy of that result as a part of their `status`. 

I agree that we are facing a memory vs speed tradeoff, but I think we get a clearer picture of this once we determine the practical number of workers on production (accompanied by benchmarking).

Please let me know your thoughts on this, I can definitely refactor this tradeoff for memory in the favor of speed in a separate PR.